common/beerocks/bwl/shared/netlink_socket.cpp

```diff
@@ -0,0 +1,147 @@
+
+namespace bwl {
+
+netlink_socket::netlink_socket(int protocol) : m_protocol(protocol)
+{
+    // Allocate the netlink socket structure
+    m_nl_socket = nl_socket_alloc();
+    if (!m_nl_socket) {
+        LOG(ERROR) << "Failed to allocate netlink socket!";
+    }
+}
+
+netlink_socket::~netlink_socket()
+{
+    // Deallocate the netlink socket structure
+    if (m_nl_socket) {
+        nl_socket_free(m_nl_socket);
+    }
+}
+
+bool netlink_socket::connect()
+{
+    // Connect the socket
+    if (nl_connect(m_nl_socket, m_protocol) != 0) {
+        LOG(ERROR) << "Failed to connect netlink socket!";
+        return false;
+    }
+
+    return true;
+}
+
+void netlink_socket::close()
+{
+    // Terminate connection and close socket
+    nl_close(m_nl_socket);
+}
+
+bool netlink_socket::send_receive_msg(std::function<bool(struct nl_msg *msg)> msg_create,
+                                      std::function<bool(struct nl_msg *msg)> msg_handle)
+{
+    // The Netlink message to send
+    std::shared_ptr<nl_msg> nl_message =
+        std::shared_ptr<struct nl_msg>(nlmsg_alloc(), [](struct nl_msg *obj) {
+            if (obj) {
+                nlmsg_free(obj);
+            }
+        });
+    LOG_IF(!nl_message, FATAL) << "Failed creating netlink message!";
+
+    // The Netlink callback set
+    std::shared_ptr<nl_cb> nl_callback =
+        std::shared_ptr<struct nl_cb>(nl_cb_alloc(NL_CB_DEFAULT), [](struct nl_cb *obj) {
+            if (obj) {
+                nl_cb_put(obj);
+            }
+        });
+    LOG_IF(!nl_callback, FATAL) << "Failed creating netlink callback!";
+
+    // Termination flag for the loop that receives the response messages. Possible values are:
+    // error == 1: initial value, request message has not be sent yet
+    // error == 0: response has been successfully received
+    // error < 0: some error occurred while receiving response
+    // Final value is used to compute the result code of this method.
+    int error = 1;
+
+    // Create standard callbacks
+    static auto nl_err_cb = [](struct sockaddr_nl *nla, struct nlmsgerr *err, void *arg) -> int {
+        int *error = (int *)arg;
+        *error     = err->error;
+        LOG(ERROR) << "Failed to process netlink message! " << strerror(*error) << " (" << *error
+                   << ")";
+        return NL_STOP;
+    };
+
+    static auto nl_finish_cb = [](struct nl_msg *msg, void *arg) -> int {
+        int *error = (int *)arg;
+        *error     = 0;
+        return NL_SKIP;
+    };
+
+    static auto nl_ack_cb = [](struct nl_msg *msg, void *arg) -> int {
+        int *error = (int *)arg;
+        *error     = 0;
+        return NL_STOP;
+    };
+
+    // Passing a lambda with capture is not supported for standard C function
+    // pointers. As a workaround, we create a static (but thread local) wrapper
+    // function that calls the capturing lambda function.
+    static __thread std::function<int(struct nl_msg * msg, void *arg)> nl_handler_cb_wrapper;
```

Comment:
Why not use the standard keyword `thread_local`? https://en.cppreference.com/w/cpp/keyword/thread_local

Reply:
Changed.